app_history.d.ts

```diff
@@ -0,0 +1,99 @@
+interface Window {
+  readonly appHistory: AppHistory;
+}
+
+interface AppHistory extends EventTarget {
+  readonly current: AppHistoryEntry|null;
+  readonly transition: AppHistoryTransition|null;
+
+  entries(): AppHistoryEntry[];
+
+  readonly canGoBack: boolean;
+  readonly canGoForward: boolean;
+
+  navigate(url: string, options?: AppHistoryNavigateOptions): Promise<void>;
+  reload(options?: AppHistoryReloadOptions): Promise<void>;
+
+  goTo(key: string, options?: AppHistoryNavigationOptions): Promise<void>;
+  back(options?: AppHistoryNavigationOptions): Promise<void>;
+  forward(options?: AppHistoryNavigationOptions): Promise<void>;
+
+  onnavigate: ((ev: Event) => any)|null;
```

Comment:
It looks like based on https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L2012 you'd want to set the `this` type here.

Also, you should use `AppHistoryNavigateEvent` for `onnavigate`, and `ErrorEvent` for `onnavigateerror`.

Finally, I'm a bit surprised by the `any` instead of `void` here, but that does seem to follow dom.d.ts conventions, so maybe it's correct. I did notice dom.d.ts uses `any` instead of `unknown` for `history.state` though, so maybe dom.d.ts is just a bit behind on best practices?

Reply:
I think it probably has to do with covariant typing, so the spec is saying I explicitly never care about the return value of this function. I might be able to get away with "unknown" or "never" there instead. "void" would mean that no function is allowed to return a value.

I went with unknown for getState() because I wanted people to cast their states. "any" would probably be more TS-standard for existing APIs, but it seems weird to me to let anys leak in like that. But I'm not 100% what the TS philosophy is on these and why.

Reply:
Yeah from what little reading I've done `unknown` seems really good for `state` and `info`. As for return types on event handlers, ¯\_(ツ)_/¯